Backend/index.js

```diff
@@ -592,3 +604,231 @@ app.post('/get-available-slots', async (req, res) => {
+    timeCounts[a] > timeCounts[b] ? a : b
+  );
+
+  return preferredPeriod;
+};
+
+
+const calculateUtility = (slot, userPreference, provider, currentTime, weights, serviceDuration, preferredPeriod) => {
+  const calculateUserPrefScore = (slot, userPreferences) => {
+    const slotTime = new Date(slot.time).toTimeString().slice(0, 5);
+    return userPreferences.includes(slotTime) ? 1 : 0;
+  }
+
+  const calculateFocusPeriodScore = (slot, provider, serviceDuration) => {
+    const dateSchedule = provider.schedule[slot.date];
+    if (!dateSchedule)
+      return 0;
+
+    const slots = dateSchedule.slots;
+    const status = dateSchedule.status;
+    const slotIndex = slots.findIndex(s => new Date(s).toTimeString().slice(0, 5) === new Date(slot.time).toTimeString().slice(0, 5));
+
+    if (slotIndex === -1 || status[slotIndex] !== 0)
+      return 0;
+
+    const requiredSlots = Math.ceil(serviceDuration * 60 / 30);
+
+    for (let i = 0; i < requiredSlots; i++) {
+      if (slotIndex + i >= status.length || status[slotIndex + i] !== 0) {
+        return 0;
+      }
+    }
+
+    const startFocusSlot = slotIndex - 1;
+    const endFocusSlot = slotIndex + requiredSlots;
+
+    // Check if there are booked slots immediately before or after the current slot
+    const beforeFocus = startFocusSlot >= 0 && status[startFocusSlot] === 1;
+    const afterFocus = endFocusSlot < status.length && status[endFocusSlot] === 1;
+
+    return beforeFocus || afterFocus ? 1 : 0;
+  };
+
+  const calculatePreferredTimeScore = (slot, preferredPeriod) => {
+    const hour = new Date(slot.time).getHours();
+    if (preferredPeriod === 'morning' && hour >= 8 && hour < 12)
+      return 1;
+    if (preferredPeriod === 'afternoon' && hour >= 12 && hour < 16)
+      return 1;
+    if (preferredPeriod === 'evening' && hour >= 16 && hour < 22)
+      return 1;
+    return 0;
+  };
+
+  //higher the proximity score, closer slot is to current time
+  const calculateProximityScore = (slot, currentTime) => {
+    const slotTime = new Date(slot.time);
+    const currentDate = currentTime.toISOString().split('T')[0]; // Current date in YYYY-MM-DD format
+    const slotDate = slotTime.toISOString().split('T')[0]; // Slot date in YYYY-MM-DD format
+
+    const dateDiff = (new Date(slotDate) - new Date(currentDate)) / (24 * 60 * 60 * 1000); // converts result to days
+    const timeDiff = Math.abs(slotTime.getTime() - currentTime.getTime()) / (60 * 60 * 1000); // converts results to hours
+
+    const totalTimeDiff = (Math.abs(dateDiff) * 24) + timeDiff; // Convert dateDiff to hours and add timeDiffInSameDay and combines them to get total time diff
+
+    const maxTimeDiff = 7 * 24; // 168 (Maximum difference is 7 days (1 week) in hours)
+
+    return 1 - (totalTimeDiff / maxTimeDiff);
+  };
+
+  const calculatePopularityScore = (slot) => {
+    return 1 - slot.popularity;
+  };
+
+  const userPrefScore = calculateUserPrefScore(slot, userPreference);
```

Comment:
Great job on this super clean code, Sewa! Love how you broke down each factor into it's own function. This makes everything super easy to follow

As a follow-up, let's actually refactor the file where these functions live in. Currently they're all in index.js, which is mainly for routing logic, it would be cleaner if we had another file that's just for this recommendation logic